iree/compiler/Conversion/CodegenUtils/ForOpCanonicalization.cpp

```diff
@@ -0,0 +1,160 @@
+//  %24 = vector.shape_cast %22 : vector<1x4xf32> to vector<4xf32>
+// ```
+// ->
+// ```
+// %22 = scf.for %arg3 = %c0 to %c4096 step %c4 iter_args(%arg4 = %20)
+//    -> vector<4xf32> {
+//    [...]
+//    scf.yield %108 : vector<4xf32>
+//  }
+// ```
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+class ForOpArgFolding final : public OpRewritePattern<scf::ForOp> {
+ public:
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+  Value FoldCarryDep(scf::ForOp forOp, Operation* op, Operation* source,
+                     Value argValue) const {
+    if (auto shapeCast = dyn_cast<vector::ShapeCastOp>(op)) {
+      if (auto souceOp = dyn_cast<vector::ShapeCastOp>(source)) {
+        if (shapeCast.getType() == souceOp.source().getType())
+          return souceOp.source();
+      }
+    } else if (auto extractOp = dyn_cast<vector::ExtractOp>(op)) {
+      if (auto broadcastOp = dyn_cast<vector::BroadcastOp>(source)) {
+        if (extractOp.getType() == broadcastOp.getSourceType())
+          return broadcastOp.source();
+      }
+    }
+    return Value();
+  }
+
+  void transferBody(Block* source, Block* dest, ArrayRef<Value> results,
+                    PatternRewriter& rewriter) const {
+    // Move all operations to the destination block.
+    rewriter.mergeBlocks(source, dest, dest->getArguments());
+    // Replace the yield op by one that returns only the used values.
+    auto yieldOp = cast<scf::YieldOp>(dest->getTerminator());
+    rewriter.updateRootInPlace(
+        yieldOp, [&]() { yieldOp.getOperation()->setOperands(results); });
+  }
+
+  LogicalResult matchAndRewrite(scf::ForOp forOp,
+                                PatternRewriter& rewriter) const override {
+    SmallVector<unsigned, 8> iteratorFolded;
+    SmallVector<Operation*, 8> resultOps;
+    auto terminator = cast<scf::YieldOp>(forOp.getBody()->getTerminator());
+    SmallVector<Value, 8> returnValues(terminator.getOperands().begin(),
+                                       terminator.getOperands().end());
+    SmallVector<Value, 8> initArgs(forOp.getIterOperands().begin(),
+                                   forOp.getIterOperands().end());
+    for (auto it : llvm::enumerate(forOp.getRegionIterArgs())) {
+      if (!it.value().hasOneUse()) continue;
+      Operation* op = it.value().use_begin()->getOwner();
+      if (!isa<vector::ShapeCastOp, vector::ExtractOp>(op)) continue;
+      Operation* returnValDef = returnValues[it.index()].getDefiningOp();
+      Value newReturn =
+          FoldCarryDep(forOp, op, returnValDef, initArgs[it.index()]);
+      if (newReturn) {
+        iteratorFolded.push_back(it.index());
+        resultOps.push_back(returnValDef);
+        returnValues[it.index()] = newReturn;
+
+        BlockAndValueMapping mapping;
+        mapping.map(op->getOperand(0), initArgs[it.index()]);
+        initArgs[it.index()] = rewriter.clone(*op, mapping)->getResult(0);
+      }
+    }
+    if (iteratorFolded.empty()) return failure();
```

Comment:
This should return success?

Reply:
Changed it, however I'm a bit confused about this one. Are we supposed to return success even if nothing is changed?

Reply:
Returning false will stop the compilation pipeline. IIUC, there are cases where the input just does not have such IRs so the pass should do nothing and exit?